sdk/dotnet/Pulumi/Config.cs

```diff
@@ -0,0 +1,361 @@
+// Copyright 2016-2018, Pulumi Corporation
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Config is a bag of related configuration state.  Each bag contains any number of
+    /// configuration variables, indexed by simple keys, and each has a name that uniquely
+    /// identifies it; two bags with different names do not share values for variables that
+    /// otherwise share the same key.  For example, a bag whose name is <c>pulumi:foo</c>, with keys <c>a</c>,
+    /// <c>b</c>, and <c>c</c>, is entirely separate from a bag whose name is <c>pulumi:bar</c> with
+    /// the same simple key names.  Each key has a fully qualified names, such as
+    /// <c>pulumi:foo:a</c>, ..., and <c>pulumi:bar:a</c>, respectively.
+    /// </summary>
+    public partial class Config
+    {
+        /// <summary>
+        /// name is the configuration bag's logical name and uniquely identifies it.  The default
+        /// is the name of the current project.
+        /// </summary>
+        public readonly string Name;
+
+        public Config(string? name = null)
+        {
+            if (name == null)
+            {
+                name = Deployment.Instance.Options.Project;
+            }
+
+            if (name.EndsWith(":config"))
+            {
+                name = name[0..^":config".Length];
+            }
+
+            this.Name = name;
+        }
+
+        private static Output<T> MakeSecret<T>(T value)
+        {
+            return new Output<T>(
+                ImmutableHashSet<Resource>.Empty,
+                Task.FromResult(new OutputData<T>(value, isKnown: true, isSecret: true)));
+        }
+
+        /// <summary>
+        /// Loads an optional configuration value by its key, or <see langword="null"/> if it doesn't exist.
+        /// </summary>
+        public string? Get(string key, StringOptions? opts = null)
```

Comment:
Super nitpick: I know we use the name `opts` pervasively in TypeScript. However, I am wondering if spelling it out as `options` for .NET would be a bit more idiomatic naming wise?

Reply:
i went back and forth on this (same with 'args' versus 'arguments'). Not certain where i land currently.

Reply:
Note that there is precedent for `args` in .NET (e.g. `EventArgs`), but not sure about `opts`.

Reply:
Yup.  I've switched it all to 'options'. Thanks!